peer: only send pings if the connection is idle for one minute

yyforyongyu opened the pull request:
Based on the [specs](https://github.com/lightning/bolts/blob/master/01-messaging.md#the-ping-and-pong-messages), the `ping` is used for keeping the liveness of the connection. We perform this check every minute, regardless of whether the connection is live or not.

This PR now changes the behavior such that we only perform this check when the connection is idle - if we are busy receiving gossip messages over the wire, we know for sure the connection is alive and can delay sending the pings.

yyforyongyu committed:
peer: reset ping ticker upon receiving new messages

This commit changes the ping/pong mechanism used for liveness check.
Previously we will always fire a ping every minute, even we know for
sure the connection is alive as we are busy processing other messages
like gossip updates. We now only fire a ping if we haven't received any
messages from the peer after one minute.

yyforyongyu committed:
docs: update release notes

morehouse reviewed (approved):
LGTM

morehouse reviewed (changes requested):
Actually there might be a slight problem here...

Currently `writeHandler` will force a disconnect if no message has been successfully sent to the peer in 5 minutes.  Prior to this change, pings would be sent every 1 minute, thus preventing this disconnect.  Now it is possible that the peer sends a gossip message often enough to prevent the pings, and then gets disconnected as a result.

Roasbeef commented:
> Prior to this change, pings would be sent every 1 minute, thus preventing this disconnect. Now it is possible that the peer sends a gossip message often enough to prevent the pings, and then gets disconnected as a result.

One idea would be to do an explicit ping check at that point before D/C. If it works (pong received), we continue onwards, otherwise, we disconnect.

Roasbeef commented:
Here's a patch that implements that idea above: https://github.com/Roasbeef/lnd/commit/120375e720fa5f349fa87a2a9eae4263e6dbbd23

Before we would: "disconnect if we haven't written a message in 5 minutes". 

Now we'll: "force a ping to be sent if we haven't written a message in 5 minutes". 

Returning to @morehouse's scenario above, instead of never disconnect, we'll send a ping, then if that doesn't receive a timely response, we'll disconnect.

ziggie1984 commented:
Really nice idea with the combination of the patch +1 on this approach.

yyforyongyu committed:
peer: delay pings on successful writes

Similar to how we delay sending pings on successful reads, we now delay
sending pings if we know the connection is up and alive for writes.

yyforyongyu committed:
peer: remove `idleTimer` in `readHandler`

Remove this `idleTimer` as the `readNextMessage` will time out after 5
seconds anyway, which means it's unlikely the `idleTimer` will fire due
to a congestion in the read pipe of the connection. When a timeout
happens, this error is not caught and will break the loop and jump
directly to the disconnection and this commit doesn't change that
behaivor.

This commit will change the behaivor that, if any of the message
processing takes longer than 5 minute, it will not disconnect but wait
for that process to be finished. Since all the processing are async,
this should never happen.

yyforyongyu committed:
peer: replace `idleTimer` with `flushTimer` in `writeHandler`

This commit removes the `idleTimer` used in the `writeHandler`. When the
connection is idle for one minute, a ping will be written to the
connection, which means the `idleTimer` won't be fired. If retrying the
write fails due to timeout, we will disconnect the peer.

yyforyongyu commented on peer/brontide.go:
```diff
@@ -2611,20 +2613,19 @@ func (p *Brontide) writeMessage(msg lnwire.Message) error {
 		select {
 		case outMsg := <-p.sendQueue:
+			// Reset the timeout.
+			flushTimer.Reset(writeFlushTimeout)
```
Given we are using go1.23, there's no need to call `Stop` anymore, [reference](https://antonz.org/go-1-23/#timer-changes).

yyforyongyu commented:
@morehouse Thanks for the detailed analysis! Yes unfortunately the remote peer can hold the ACK since TCP is full-duplex so they can keep sending us msgs without ACKing ours. Will put this PR on hold and revisit for 0.20, also have some ideas about how to better design this area, will create a doc to kick off the discussions.

lightninglabs-deploy commented:
@yyforyongyu, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@yyforyongyu, remember to re-request review from reviewers when ready

yyforyongyu commented:
!lightninglabs-deploy mute
